tokio/src/sync/mpsc/chan.rs

```diff
@@ -336,6 +345,21 @@ where
             }
         })
     }
+
+    /// Returns `true` if the channel is closed.
+    ///
+    /// The channel is closed if the number of outstanding sender
+    /// handles is zero, thus the send half of the channel is closed,
+    /// or if `Rx::close` is called.
+    pub(crate) fn is_closed(&self) -> bool {
+        if self.inner.tx_count.load(SeqCst) == 0 {
```

Comment:
Currently, "poll_ready` returning "closed" indicates that there are no senders left (ref count == 0) **and** all messages have been consumed. With this implementation, it would be possible for `is_closed()` to return `true` but there are still pending messages to receive.

My guess is the correct implementation will be very similar to the `try_recv` path except it would a) not consume a message and b) return `true` instead of `TryRecvError::Closed`.

There is also an API issue to work through. `Receiver::close` signals to the sender that the receiver wishes to stop receiving, but `Receiver::is_closed()` would indicate that all senders dropped and all messages have been consumed.

We may want to differentiate with `is_send_closed()` or something like that. Thoughts?

Reply:
> Currently, "poll_ready`returning "closed" indicates that there are no senders left (ref count == 0) **and** all messages have been consumed. With this implementation, it would be possible for`is_closed()`to return`true` but there are still pending messages to receive.
> 
> ... the correct implementation will be very similar to the `try_recv` path except it would a) not consume a message and b) return `true` instead of `TryRecvError::Closed`.

I see; yes. It surely makes sense.

> There is also an API issue to work through. `Receiver::close` signals to the sender that the receiver wishes to stop receiving, but `Receiver::is_closed()` would indicate that all senders dropped and all messages have been consumed.
> 
> We may want to differentiate with `is_send_closed()` or something like that. Thoughts?

It's indeed an important distinction; I'll work on that.